src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs

```diff
@@ -100,25 +125,76 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
                 while (userName.Length == 0);
             }
 
-            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName
-            );
+            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName);
 
-            //
             // now, prompt for the password
-            //
-            WriteToConsole(passwordPrompt, true);
-            password = ReadLineAsSecureString();
-            if (password == null)
+            do
             {
-                return null;
+                WriteToConsole(passwordPrompt, true);
+                password = ReadLineAsSecureString();
+                if (password == null)
+                {
+                    return null;
+                }
             }
+            while (password.Length == 0);
 
-            WriteLineToConsole();
+            if (reenterPassword)
+            {
+                reenterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ReenterPassword, userName);
+                passwordMismatch = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_PasswordMismatch);
+                do
+                {
+                    // now, prompt to re-enter the password.
+                    WriteToConsole(reenterPasswordPrompt, true);
+                    confirmPassword = ReadLineAsSecureString();
+                    if (confirmPassword == null)
+                    {
+                        return null;
+                    }
 
-            cred = new PSCredential(userName, password);
+                    if (SecureStringEquals(password, confirmPassword))
+                    {
+                        break;
+                    }
 
+                    WriteToConsole(ConsoleColor.Red, ConsoleColor.Black, passwordMismatch, true);
+                }
+                while (true);
+            }
+
+            WriteLineToConsole();
+            cred = new PSCredential(userName, password);
             return cred;
         }
+
+        private static bool SecureStringEquals(SecureString password, SecureString confirmPassword)
+        {
+            IntPtr pwd_ptr = IntPtr.Zero;
+            IntPtr confirmPwd_ptr = IntPtr.Zero;
+            try
+            {
+                pwd_ptr = Marshal.SecureStringToBSTR(password);
+                string pwd = Marshal.PtrToStringBSTR(pwd_ptr);
```

Comment:
Converting passwords to plain text, leaves multiple immutable string objects in memory, indefinitely.  I think it would be better to convert to byte array for comparison, and the byte array values can be zeroed out afterwards.  Also you can compare byte array length for early out.

Reply:
@PaulHigin Thank you for the review. I have updated the function to use byte array for comparison and have zeroed it in the finally clause. The memory space where the pointer was pointing had a null value between each byte so I had to double the length to use as size. Please review and provide your feedback.